Late escape Social block

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
This is not a security problem. This PR simply moves [escaping of all PHP output to be as "late" as possible](https://developer.wordpress.org/themes/theme-security/data-sanitization-escaping/#:~:text=esc_textarea(%20%24text%20)%3B%20%3F%3E%3C/textarea%3E-,Tip%3A,Output%20escaping%20should%20occur%20as%20late%20as%20possible.,-Top%20%E2%86%91). This means we avoid escaping variables until they are output in the HTML markup.

This is a WP Core best practice.

## How has this been tested?
- add blocks
- check functionality is "as was"
- check all tests pass

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

getdave committed:
First pass social link block

aristath commented on packages/block-library/src/social-link/index.php:
```diff
@@ -32,9 +32,9 @@ function render_block_core_social_link( $attributes, $content, $block ) {
+	$rel_target_attributes = '';
 	if ( $open_in_new_tab ) {
-		$attribute = 'rel="noopener nofollow" target="_blank"';
+		$rel_target_attributes = 'rel="' . esc_attr( 'noopener nofollow' ) . '" target="' . esc_attr( '_blank' ) . '"';
```
Since these are hardcoded and there's no user-input involved, there's no reason to escape them.

getdave replied:
Yeh I take your point.

To give you an idea of my thinking, I'm trying to avoid situations where we have hard coded variables that are not escaped at point of output.

Whilst these variables might contain hardcoded content _now_ there's nothing to say they will stay that way. What if developer A comes along and makes the `rel` attribute configurable from the block. What if they simply put the attribute provided by the user into the variable above and forget to escape? What happens then?

By being a little more verbose with escaping now, we avoid potential future XSS issues being introduced. For example if Developer A were to come along to the code I've suggested here there's no danger they will forget to escape the `rel` attribute. Personally I think the trade off is worth it.

This is a unusual use case because we aren't actually escaping as close to the building of the HTML as I"d like due to the conditional rendering of the x2 attributes. Ideally we'd simply escape the output in the HTML template rather than do it when we're building the variable.

Don't get me wrong, I understand this is pretty verbose and might seem like over doing things. However in my experience it's better to be safe than sorry.

getdave replied:
If you're not happy with this we can just close this PR out.

Mamaduka replied:
I'm with Ari on this one. There's no need to escape hardcoded strings.

getdave replied:
Actually reflecting on this again, I've now _removed_ the variable and so these strings are inline. Therefore I agree why are we escaping? 

My main concern is variables not being escaped. I'll make the change.

Mamaduka replied:
Thanks, Dave.

getdave replied:
When I started making the change I noticed that the `rel` attribute is in fact _not_ inlined. I still have reservations but they are relatively minor so I've removed the escaping as per the majority opinion.

Mamaduka commented on packages/block-library/src/social-link/index.php:
```diff
@@ -45,7 +45,7 @@ function render_block_core_social_link( $attributes, $content, $block ) {
 	);
 
-	return '<li ' . $wrapper_attributes . '><a href="' . esc_url( $url ) . '" aria-label="' . esc_attr( $label ) . '" ' . $attribute . ' class="wp-block-social-link-anchor"> ' . $icon . '</a></li>';
+	return '<li ' . $wrapper_attributes . '><a href="' . esc_url( $url ) . '" aria-label="' . esc_attr( $label ) . '" ' . $rel_target_attributes . ' class="wp-block-social-link-anchor"> ' . esc_html( $icon ) . '</a></li>';
```
Escaping icon breaks the block output since the icons are SVGs.


![CleanShot 2022-01-14 at 11 36 59](https://user-images.githubusercontent.com/240569/149469325-ae00cba2-5657-41e5-a1ca-671619fbfabd.png)

getdave replied:
Thank you for catching this 😌 . I've been going through trying to remove instances where I'd inadvertently escaped SVGs.

getdave replied:
Fixed.

getdave committed:
Remove escaping of hard coded strings

Addresses https://github.com/WordPress/gutenberg/pull/37881#discussion_r782942702

getdave committed:
Remove escaping of SVG

Addresses https://github.com/WordPress/gutenberg/pull/37881#pullrequestreview-852606370

getdave commented:
This PR now becomes "renaming a variable" and "removing a space character" so I'm not fussed about merging it. I guess we could if folks feel it's worth while 😄
